pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/SegmentMergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,205 @@
+              DEFAULT_MAX_NUM_SEGMENTS_PER_TASK);
+
+      int maxNumRecordsPerSegment =
+          readIntConfigWithDefaultValue(taskConfigs, MinionConstants.MergeRollupTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+              DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT);
+
+      String bufferTimePeriodStr =
+          taskConfigs.getOrDefault(MinionConstants.MergeRollupTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_TIME_PERIOD);
+      long bufferTimePeriodInMillis;
+      try {
+        bufferTimePeriodInMillis = TimeUtils.convertPeriodToMillis(bufferTimePeriodStr);
+      } catch (IllegalArgumentException e) {
+        LOGGER.warn(
+            "Buffer time period ('{}') for table '{}' is not configured correctly. Falling back to default behavior ('{}')",
+            bufferTimePeriodStr, offlineTableName, DEFAULT_BUFFER_TIME_PERIOD);
+        bufferTimePeriodInMillis = TimeUtils.convertPeriodToMillis(DEFAULT_BUFFER_TIME_PERIOD);
+      }
+
+      // Generate tasks
+      int tableNumTasks = 0;
+
+      List<OfflineSegmentZKMetadata> segmentsForOfflineTable =
+          _clusterInfoProvider.getOfflineSegmentsMetadata(offlineTableName);
+
+      // Fetch the segment lineage for the table and compute the segments that should not be scheduled for merge
+      // based on the segment lineage.
+      SegmentLineage segmentLineageForTable = _clusterInfoProvider.getSegmentLineage(offlineTableName);
+      Set<String> segmentsNotToMerge = new HashSet<>();
+      if (segmentLineageForTable != null) {
+        for (String segmentLineageEntryId : segmentLineageForTable.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineageForTable.getLineageEntry(segmentLineageEntryId);
+          // Segments shows up on "segmentFrom" field in the lineage entry should not be scheduled again.
+          segmentsNotToMerge.addAll(lineageEntry.getSegmentsFrom());
+
+          // Segments shows up on "segmentsTo" field in the lineage entry with "IN_PROGRESS" state cannot be merged.
+          if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+            segmentsNotToMerge.addAll(lineageEntry.getSegmentsTo());
+          }
+        }
+      }
+
+      // Filter out the segments that cannot be merged
+      List<SegmentZKMetadata> segmentsToMergeForTable = new ArrayList<>();
+      List<String> scheduledSegments = scheduledSegmentsMap.getOrDefault(offlineTableName, Collections.emptyList());
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : segmentsForOfflineTable) {
+        String segmentName = offlineSegmentZKMetadata.getSegmentName();
+
+        // The segment should not be merged if it's already scheduled or in progress
+        if (scheduledSegments.contains(segmentName) || segmentsNotToMerge.contains(segmentName)) {
+          continue;
+        }
+
+        // The segment should not be merged if it already contains enough records
```

Comment:
This is a good point since rollup will reduce the number of rows greatly. I will just let `SegmentProcessorFramework` to deal with the number of records. I will remove this check.